colossalai/pipeline/schedule/zero_bubble_pp.py

```diff
@@ -472,19 +478,47 @@ def backward_b_step(
         # calculate bwd b step ; only dx = w*dy;
 
         # Retain the grad on the input_obj.
-        tree_map(retain_grad, input_obj)
+        if input_obj is None:
+            return None
+        else:
+            tree_map(retain_grad, input_obj)
+
+        # x, y, dy list for backward_by_grad; Type: list[tensor];
+        input_obj_ = []
+        output_obj_ = []
+        output_obj_grad_ = []
+
+        # get x from input_obj to input_obj_
+        for k, v in input_obj.items():
+            if v.requires_grad:
+                input_obj_.append(input_obj[k])
 
         if model_chunk_id == 1 and self.stage_manager.is_first_stage(ignore_chunk=True):
             # loss backward; output_obj is loss; so output_obj_grad should be None
             assert output_obj_grad is None
+            output_obj_grad_.append(output_obj_grad)  # None
+            output_obj_.append(output_obj)  # LOSS
+
+        else:
+            for k, v in input_obj.items():
+                if v.requires_grad:
+                    output_obj_.append(output_obj[k])
+                    output_obj_grad_.append(output_obj_grad[k])
 
         optimizer.backward_by_grad(
-            tensor=output_obj,
-            grad=output_obj_grad,
-            inputs=input_obj,
+            tensor=output_obj_,
+            grad=output_obj_grad_,
+            inputs=input_obj_,
             retain_graph=True,
         )
-        return input_obj.grad
+
+        # format output_obj_grad
+        if input_obj is not None:
+            input_obj_grad = {}
+            for k, v in input_obj.items():
+                if isinstance(v, torch.Tensor) and v.grad is not None:
+                    input_obj_grad[k] = v.grad
```

Comment:
should use tree_map or tree_flatten

Reply:
> should use tree_map or tree_flatten

Fixed in https://github.com/hpcaitech/ColossalAI/pull/6065/commits/c6d6ee39bda6d0e3aa0a6233796a0d1059eb30dc;